pkg/operator/vspherecontroller/driver_starter.go

```diff
@@ -324,3 +330,65 @@ func getOperatorSyncState(operatorClient v1helpers.OperatorClientWithFinalizers)
 	}
 	return opSpec.ManagementState
 }
+
+// setMaxVolumesPerNodeEnvVar sets the MAX_VOLUMES_PER_NODE environment variable
+// in container specifications based on ClusterCSIDriver configuration.
+func setMaxVolumesPerNodeEnvVar(containers []v1.Container, clusterCSIDriverLister operatorlistersv1.ClusterCSIDriverLister, featureGate featuregates.FeatureGate) ([]v1.Container, error) {
+	// Get ClusterCSIDriver object
+	clusterCSIDriver, err := clusterCSIDriverLister.Get(utils.VSphereDriverName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ClusterCSIDriver: %v", err)
+	}
+
+	// Get the maximum volume limit
+	maxVolumesPerNode := utils.GetMaxVolumesPerNode(clusterCSIDriver, featureGate)
+
+	// Set the environment variable in all containers
+	for i := range containers {
+		container := &containers[i]
+
+		// Find and update existing MAX_VOLUMES_PER_NODE env var if it exists
+		envVarFound := false
+		for j := range container.Env {
+			if container.Env[j].Name == "MAX_VOLUMES_PER_NODE" {
+				container.Env[j].Value = strconv.FormatInt(int64(maxVolumesPerNode), 10)
+				envVarFound = true
+				break
+			}
+		}
+
+		// If not found, append it
+		if !envVarFound {
+			container.Env = append(container.Env, v1.EnvVar{
+				Name:  "MAX_VOLUMES_PER_NODE",
+				Value: strconv.FormatInt(int64(maxVolumesPerNode), 10),
+			})
+		}
+	}
+
+	return containers, nil
+}
+
+// WithMaxVolumesPerNodeDaemonSetHook sets the MAX_VOLUMES_PER_NODE environment variable in the DaemonSet container specifications.
+func WithMaxVolumesPerNodeDaemonSetHook(clusterCSIDriverLister operatorlistersv1.ClusterCSIDriverLister, featureGate featuregates.FeatureGate) csidrivernodeservicecontroller.DaemonSetHookFunc {
+	return func(opSpec *operatorapi.OperatorSpec, ds *appsv1.DaemonSet) error {
+		containers, err := setMaxVolumesPerNodeEnvVar(ds.Spec.Template.Spec.Containers, clusterCSIDriverLister, featureGate)
+		if err != nil {
+			return err
+		}
+		ds.Spec.Template.Spec.Containers = containers
+		return nil
+	}
+}
+
+// WithMaxVolumesPerNodeDeploymentHook sets the MAX_VOLUMES_PER_NODE environment variable in the Deployment container specifications.
+func WithMaxVolumesPerNodeDeploymentHook(clusterCSIDriverLister operatorlistersv1.ClusterCSIDriverLister, featureGate featuregates.FeatureGate) deploymentcontroller.DeploymentHookFunc {
```

Comment:
I didn't think this was required in control-plane deployment.

Reply:
That's what I thought as well but it was added to assets here: https://github.com/openshift/vmware-vsphere-csi-driver-operator/commit/7cf46b969cc1329e9d09b855747ee0d1f9bc6a76

And `CSINode` is created for master nodes too which is where the volume limit is reflected as:
```
spec:
  drivers:
  - allocatable:
      count: 59
```

But the reasons are not entirely clear to me, maybe consistency? Or because we can't prevent workloads being scheduled on masters?

Anyway, if you think this should be removed we shouldn't do it in scope of this PR as it would be unrelated change.

Reply:
The allocatable limit from control-plane drivers don't make any sense. The master nodes actually should have 2 drivers running. One in worker mode and another one in control-plane mode. The driver that runs with controller doesn't have any impact on reported limit.

To be clear - I am *not* asking you to remove the limit from asset file. I am just asking you to not add this hook at all here. Just leave it as it is. Less code the better.

Reply:
Alright, removed.